src/Raven.Server/Documents/DocumentsStorage.cs

```diff
@@ -1564,28 +1565,38 @@ public void AssertMetadataKey(string id, BlittableJsonReaderObject document, Doc
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private static Document ParseDocument(JsonOperationContext context, ref TableValueReader tvr, DocumentFields fields)
+        private static Document ParseDocument(DocumentsOperationContext context, ref TableValueReader tvr, DocumentFields fields)
         {
             if (fields == DocumentFields.All)
             {
-                return new Document
+                var document = new Document
                 {
                     StorageId = tvr.Id,
                     LowerId = TableValueToString(context, (int)DocumentsTable.LowerId, ref tvr),
                     Id = TableValueToId(context, (int)DocumentsTable.Id, ref tvr),
                     Etag = TableValueToEtag((int)DocumentsTable.Etag, ref tvr),
-                    Data = new BlittableJsonReaderObject(tvr.Read((int)DocumentsTable.Data, out int size), size, context),
                     ChangeVector = TableValueToChangeVector(context, (int)DocumentsTable.ChangeVector, ref tvr),
                     LastModified = TableValueToDateTime((int)DocumentsTable.LastModified, ref tvr),
                     Flags = TableValueToFlags((int)DocumentsTable.Flags, ref tvr),
                     TransactionMarker = TableValueToShort((int)DocumentsTable.TransactionMarker, nameof(DocumentsTable.TransactionMarker), ref tvr),
                 };
+
+                if (context.Transaction.InnerTransaction.IsCompressed(tvr.Id) == false)
+                {
+                    document.Data = new BlittableJsonReaderObject(tvr.Read((int)DocumentsTable.Data, out int size), size, context);
+                    return document;
+                }
+                else
+                {
+                    document.Data = new CompressedBlittableJsonReaderObject(tvr.Read((int)DocumentsTable.Data, out int size), size, context, tvr.Id);
+                    return document;
+                }
```

Comment:
I would make a method of `CreateBlittableJsonReaderObject` and reuse it

Reply:
removed this check